### PR TITLE
persist: understand structure of persisted data

### PR DESCRIPTION
**Pull request by danhhz**
For efficiency/performance, we directly expose the columnar structure of persist's internal encoding to users during encoding and decoding. For ergonomics, we wrap the arrow2 crate we use to read and write parquet data.

Some of the requirements that led to this design:
- Support a separation of data and schema because Row is not self-describing: e.g. a Datum::Null can be one of many possible column types. A RelationDesc is necessary to describe a Row schema.
- Narrow down arrow2::data_types::DataType (the arrow "logical" types) to a set we want to support in persist.
- Associate an arrow2::io::parquet::write::Encoding with each of those types.
- Do `dyn Any` downcasting of columns once per part, not once per update.
- Unlike arrow2, be precise about whether each column is optional or not.

The primary presentation of this abstraction is a sealed trait [Data], which is implemented for the owned version of each type of data that can be stored in persist: `int64`, `Option<String>`, etc.

Under the hood, it's necessary to store something like a map of `name -> column`. A natural instinct is to make Data object safe, but I couldn't figure out a way to make that work without severe limitations. As a result, the DataType enum is introduced with a 1:1 relationship between variants and implementations of Data. This allows for easy type erasure and guardrails when downcasting the types back.

Note: The "Data" strategy is roughly how columnation works and the "DataType" strategy is roughly how arrow2 works. Doing both of them gets us the benefits of both, while the downside is code duplication and cognitive overhead.

The Data trait has associated types for the exclusive "builder" type for the column and for the shared "reader" type. These implement also implement some common traits to make relationships between types more structured.

Finally, the [Schema] trait maps an implementor of [Codec] to the underlying column structure. It also provides a [PartEncoder] and [PartDecoder] for amortizing any downcasting that does need to happen.

Initial benchmarks compared to the previous Protobuf based encoding for Row (specifically, 1000 of them each with one bool and one string):

    roundtrip_encode_legacy time:   [1.5251 ms 1.5267 ms 1.5285 ms]
    roundtrip_encode_structured
                            time:   [852.47 µs 852.99 µs 853.65 µs]
    roundtrip_decode_legacy time:   [1.9244 ms 1.9314 ms 1.9376 ms]
    roundtrip_decode_structured
                            time:   [619.88 µs 620.12 µs 620.40 µs]

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

Rough timeline I'm imagining:

- Hash out this PR and get it merged with the code completely unused.
- In a separate PR, figure out the full mapping for SourceData -> parquet. Hook this up in some sort of checks mode (ideally in CI, maybe behind soft assert?) to sanity check that data we see in practice does in fact faithfully roundtrip.
- In a final PR, turn this on an introduce a migration codepath to handle existing pre-strutured data.

There's quite a bit of boilerplate in row/encoding.rs. I don't think it's fundamental (seems like DatumType and AsColumnType already most of the work once we can talk rust into it), but I'm hesitant to spend too much time on that before we figure out the full SourceData -> parquet mapping.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by danhhz** [comment]
Okay @bkirwi and @pH14 this should be ready for a first round!

**Review by bkirwi — COMMENTED** [review]
Neat!

This does seem minimally invasive; I'd be happy to merge something with this shape and iterate in `main`.

File: src/persist-types/src/parquet.rs
Comment (danhhz) [reply src/persist-types/src/parquet.rs:63]: yep! I'll type a comment to that effect to make it more obvious

it actually does check that it's a single element iterator by erring if the first call to `next` is a `None` or if the second call to `next` is a `Some(_)` (see below). sounds like that was unobvious enough that it also deserves a comment

File: src/persist-types/src/part.rs
Comment (danhhz) [reply src/persist-types/src/part.rs:78]: none of this code handles any sort of schema migration or mismatch right now and that's on purpose for V0 (I assume you noticed that, but saying it just to be explicit). I think it's a good idea to start hashing out our story here, though!

eliding the field makes me a bit nervous[^1], but I imagine our rules could be something like this:

- no fields -> one optional field: decode expects a structarray with one column, but gets nullarray. synthesizes an structarray with a single all-null column
- N fields -> N fields + 1 optional field (where N > 0): the optional column is missing from the structarray, decode fills it in with a single all-null column
- N fields + 1 optional field -> N fields (where N >0): decode drops the relevant column from the structarray it receives
- one optional field -> no fields: decode ignores the relevant column

which all means encode is free to always write data using the current schema

thoughts?

[^1]: right now we're structuring these as logically `Vec<(K, V, T, D)>` but there's a world in which we might want to structure them as `Vec<(K, Vec<(V, Vec<(T, D)>)>)>`, which takes advantage of fancy dremel/parquet things and also more directly matches the [structure of in-memory spine batches](https://github.com/TimelyDataflow/differential-dataflow/blob/8a04699df22600a332a5fafadbe7eebb84ea16d5/src/trace/implementations/ord.rs#L88-L104). eliding things makes that more complicated

File: src/persist-types/src/parquet.rs
Comment (bkirwi) [reply src/persist-types/src/parquet.rs:63]: No, the Some check is clear and I managed to totally miss it! No need for extra commentary I think.

File: src/persist-types/src/part.rs
Comment (bkirwi) [reply src/persist-types/src/part.rs:78]: Yeah, definitely not expecting us to have a proper schema migration story, but nice if we can avoid writing down data in such a way that it makes future schema migration stuff harder!

The n<>n+1 bullets make total sense to me. 0<>1 makes me nervous because it's the only circumstance where we'd change the type of a parquet column, which otherwise we never do. "We never change the type of a parquet column" sounds like a really nice invariant! Also "no column" to "struct column with a single field" sounds like how we'd handle schema evolution for nested structs, and I don't see a strong reason to handle it differently at the top level.

Re: the footnote... IIUC `Vec<(K, Vec<(V, Vec<(T, D)>)>)>` translates effectively to `[{k: K, vtd: [{v: V, td: [{t: T, d: D}}]}]}]` in parquet. (Just translating the tuples into records.) A null value would then be `[{k: K, vtd: [{td: [{t: T, d: D}}]}]}]`, which seems basically fine? But if you disagree this is probably something for a sync chat instead of hashing it out here... the dremel-style record shredding does get a little gnarly.

Comment (bkirwi) [reply src/persist-types/src/part.rs:78]: (FWIW I don't think we necessarily need to nail this down before merging, since we're not actually persisting the results anywhere yet! It's just the one place where the correct encoding for a given schema isn't totally straightforward.)

Comment (danhhz) [reply src/persist-types/src/part.rs:78]: > 0<>1 makes me nervous because it's the only circumstance where we'd change the type of a parquet column, which otherwise we never do. "We never change the type of a parquet column" sounds like a really nice invariant!

Not sure I grokked the first time around that your point was that the column's type changed. As for the latter bit, I feel like "We always have K, V, T, and D columns" would also be a nice invariant! :D (That said, see below.)

> Also "no column" to "struct column with a single field" sounds like how we'd handle schema evolution for nested structs, and I don't see a strong reason to handle it differently at the top level.

This is a great point and actually far more compelling to me.

> `Vec<(K, Vec<(V, Vec<(T, D)>)>)>` translates effectively to `[{k: K, vtd: [{v: V, td: [{t: T, d: D}}]}]}]` in parquet. (Just translating the tuples into records.) A null value would then be `[{k: K, vtd: [{td: [{t: T, d: D}}]}]}]`, which seems basically fine? But if you disagree this is probably something for a sync chat instead of hashing it out here... the dremel-style record shredding does get a little gnarly.

👍 Yeah, good call. This also nicely parallels with DD, which special cases `(K, ())` with a different batch type that skips over the val layer.

Okay, I'm convinced! I would, however, like to hold off on any code changes for this until (a) Paul has had a chance to leave a first round of comments and (b) I've taken a moment to sketch out Struct columns as they'd be exposed to the persist user (just to confirm the intuition above about nested and top-level structs matching).

Comment (bkirwi) [reply src/persist-types/src/part.rs:78]: > As for the latter bit, I feel like "We always have K, V, T, and D columns" would also be a nice invariant! :D

This is fair! I think I have a latent stress about column type changes from my time as a data infrastructure person, where it's something we'd avoid ~at all costs. (Whereas since we're writing the mapping code here, we don't necessarily have the same restrictions.)

> This also nicely parallels with DD, which special cases (K, ()) with a different batch type that skips over the val layer.

Cool!

> I would, however, like to hold off on any code changes for this until [...]

Yep, sounds good, as you prefer.

File: src/persist-types/src/part.rs
Comment (pH14) [inline src/persist-types/src/part.rs:27]: dropped a note in Slack on this one, but it's not super clear to me what `Part` does versus an in-mem arrow representation

File: src/persist-types/src/part.rs
Comment (pH14) [inline src/persist-types/src/part.rs:344]: ran into this pretty quickly while playing around with things -- we want this to be `dyn Any + Send` (and the associated changes on `Data::Mut` and `Data::Col` etc) so we can use a part builder inside a future

**Review by pH14 — COMMENTED** [review]
So far making sense to me, though I'm really only able to grok this one by playing around with it. It's a bit out of my wheelhouse both in content (not familiar with the internals of arrow, parquet) and in implementation (rust details are >= my rust abilities) so I'll keep dropping comments as I work on it. Very exciting!

File: src/repr/src/row/encoding.rs
Comment (pH14) [inline src/repr/src/row/encoding.rs:673]: don't think we need to go crazy with tests right away, but we should probably have tests that roundtrip more than 1 row

**Review by pH14 — APPROVED** [review]
Been playing around with this branch and it seems like it ought to do the trick! The integration / rebase with Gus's handiwork LGTM. Do you have any specific criteria in mind to check off before merging?

File: src/repr/src/row/encoding.rs
Comment (danhhz) [reply src/repr/src/row/encoding.rs:673]: note: Paul retracted this request offline in the interest of getting this merged now

**Review by danhhz — COMMENTED** [review]
TFTRs!!

> This does seem minimally invasive; I'd be happy to merge something with this shape and iterate in main.

Hey @bkirwi, heads up that Paul has requested this get merged ASAP so he doesn't have to keep things rebased on top of it. Given this comment, I'm gonna take you up on this and hit the merge button

**Pull request merged**
